internal/mdns/mdns.go

```diff
@@ -0,0 +1,225 @@
+type Service struct {
+	server     *mdns.Server
+	p2pHost    IDNetworker
+	serviceTag string
+	notifee    Notifee
+	logger     Logger
+	started    bool
+	stop       chan struct{}
+	done       chan struct{}
+}
+
+// NewService creates and returns a new mDNS service.
+func NewService(p2pHost IDNetworker, serviceTag string,
+	logger Logger, notifee Notifee) (service *Service) {
+	if serviceTag == "" {
+		serviceTag = "_ipfs-discovery._udp"
+	}
+
+	return &Service{
+		p2pHost:    p2pHost,
+		serviceTag: serviceTag,
+		notifee:    notifee,
+		logger:     logger,
+	}
+}
+
+// Start starts the mDNS service.
+func (s *Service) Start() (err error) {
+	ips, port := getMDNSIPsAndPort(s.p2pHost)
+
+	hostID := s.p2pHost.ID()
+
+	hostIDPretty := hostID.Pretty()
+	txt := []string{hostIDPretty}
+	mdnsService, err := mdns.NewMDNSService(hostIDPretty, s.serviceTag, "", "", int(port), ips, txt)
+	if err != nil {
+		return fmt.Errorf("creating mDNS service: %w", err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: mdnsService})
+	if err != nil {
+		return fmt.Errorf("creating mDNS server: %w", err)
+	}
+	s.server = server
+
+	s.started = true
+	s.stop = make(chan struct{})
+	s.done = make(chan struct{})
+	ready := make(chan struct{})
+
+	go s.run(ready)
+	// It takes a few milliseconds to launch a goroutine
+	// so we wait for the run goroutine to be ready.
+	<-ready
+
+	return nil
+}
+
+// Stop stops the mDNS service and server.
+func (s *Service) Stop() (err error) {
+	if !s.started {
+		return nil
+	}
+
+	defer func() {
+		s.started = false
+	}()
+	close(s.stop)
+	<-s.done
+	return s.server.Shutdown()
+}
+
+func (s *Service) run(ready chan<- struct{}) {
+	defer close(s.done)
+
+	const pollPeriod = time.Minute
+	ticker := time.NewTicker(pollPeriod)
+	defer ticker.Stop()
+
+	handleEntriesReady := make(chan struct{})
+	entriesListeningLoopStop := make(chan struct{})
+	entriesListeningLoopDone := make(chan struct{})
+	entriesCh := make(chan *mdns.ServiceEntry, 16)
+	entriesStartListening := make(chan struct{})
+	entriesStopListening := make(chan struct{})
+
+	go s.handleEntries(handleEntriesReady, entriesListeningLoopStop, entriesListeningLoopDone,
+		entriesStartListening, entriesStopListening, entriesCh)
+	<-handleEntriesReady
+
+	const queryTimeout = 5 * time.Second
+	params := &mdns.QueryParam{
+		Domain:  "local",
+		Entries: entriesCh,
+		Service: s.serviceTag,
+		Timeout: queryTimeout,
+	}
+
+	close(ready)
+
+	for {
+		entriesStartListening <- struct{}{}
+		err := mdns.Query(params)
+		if err != nil {
+			s.logger.Warnf("mdns query failed: %s", err)
+		}
+		entriesStopListening <- struct{}{}
+
+		// Drain the entries channel, we no longer care about entries.
+		for len(entriesCh) > 0 {
+			<-entriesCh
+		}
+
+		select {
+		case <-ticker.C:
+		case <-s.stop:
+			close(entriesListeningLoopStop)
+			<-entriesListeningLoopDone
+			close(entriesCh)
+			close(entriesStartListening)
+			close(entriesStopListening)
+			return
+		}
+	}
```

Comment:
This whole thing seems overly complicated to just periodically query the mdns and handle the entries.  Couldn't we just run a goroutine in the loop itself and just process the events for each query? I created [another PR](https://github.com/ChainSafe/gossamer/pull/2986) against this one to demonstrate.

Reply:
It was simpler (similar to your PR) before commit https://github.com/ChainSafe/gossamer/pull/2794/commits/9a0ae60f79ee59aa907faca4634741b31f5549d9

Why I changed it to this was to:
- avoid closing the channel on the reading side (our side)
- re-use the same buffered channel without creating it on every tick... but yeah the tick is 1 minute, I didn't realise :sweat: 

Reverted it to how it was (a bit better with ready and done channels)